src/Sylius/Component/Promotion/Generator/PercentageGenerationPolicy.php

```diff
@@ -68,6 +68,11 @@ private function calculatePossibleGenerationAmount(PromotionCouponGeneratorInstr
             $instruction->getSuffix()
         );
 
-        return (int) floor((16 ** $expectedCodeLength) * $this->ratio - $generatedAmount);
+        $codeCombination = 16 ** $expectedCodeLength;
```

Comment:
What I'm thinking about it is running it in the loop:

```php
for ($i = 1; $i <= $expectedCodeLength; ++$i) {
    $possibleGenerationAmount = (int) floor((16 ** $i) * $this->ratio - $generatedAmount);
    
    if ($possibleGenerationAmount >= $expectedGenerationAmount) {
        return true;
    }
}

return false;
```

It's a rough idea, probably should be simplified not to calculate `16 ** $i` every time but to multiply it by 16 on every iteration.

Reply:
We can also improve the calculation algorithm to use smaller numbers:

`(16 ** $expectedCodeLength) * $this->ratio` can become `(16 ** ($expectedCodeLength - 1)) * (16 * $this->ratio)`